### PR TITLE
[Backtracing] Add a caching wrapper for MemoryReader.

### PR DESCRIPTION
**Pull request by al45tair**
Currently we read many small chunks from the process we're backtracing. If it so happens that it's the local process, that isn't really a big problem, but if it's a remote process, especially on Linux where we have to use the memory server, it's probably a little slow.

Fix by adding a caching layer.

rdar://117681625

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
This will conflict with https://github.com/apple/swift/pull/69508; we'll need to update (probably this one) after that one lands.

File: stdlib/public/Backtracing/CachingMemoryReader.swift [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    self.cache = [:]
+  }
+
+  deinit {
+    for (_, page) in cache {
+      page.deallocate()
+    }
+  }
+
+  private func getPage(at address: Address) throws -> UnsafeRawBufferPointer {
```
Comment (mikeash) [inline stdlib/public/Backtracing/CachingMemoryReader.swift:43]: I sense that a `precondition(address & pageMask == 0)` will save someone some trouble in about four years.

**Review by mikeash — APPROVED** [review]

File: stdlib/public/Backtracing/CachingMemoryReader.swift (hunk 1 above)
Comment (al45tair) [reply stdlib/public/Backtracing/CachingMemoryReader.swift:43]: :-D Good idea.

**Comment by al45tair** [comment]
Hah. Build failed because of something I'd already fixed but failed to push. Let's try that again.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
Rebased again.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Pull request merged**